File: ruby/lib/xi/ml/tools/logger.rb
# encoding: utf-8



# The predefined and updatable logger
class Xi::ML::Tools::Logger

  ROOT = 'xi::ml'.freeze

  # Create a new logger
  #
  # @param name [String] the logger's class name
  # @return [Log4r::Logger] return the newly created logger
  def self.create(name='')
    create_root() if Log4r::Logger[ROOT].nil?

    unless Log4r::Logger[name]
      logger = Log4r::Logger.new(name)
      logger.additive = false
      logger.level = Log4r::Logger[ROOT].level
      logger.outputters = Log4r::Logger[ROOT].outputters.clone
    end

    Log4r::Logger[name]
  end

  # Change the logging level of all existing loggers
  #
  # @param level [Integer] the new logger level value
  def self.global_level=(level)
    Log4r::Logger.each do |lname|
      Log4r::Logger[lname].level = level if lname.start_with?(ROOT)
    end
  end

  # Copy the configuration of a reference logger onto all existing loggers
  #
  # @param logger [Log4r::Logger] the reference logger
  def self.copy_config(logger)
    return unless logger.is_a?(Log4r::Logger)
    Log4r::Logger.each do |lname|
      if lname.start_with?(ROOT)
        Log4r::Logger[lname].level = logger.level
        Log4r::Logger[lname].outputters = logger.outputters.clone
      end
    end
  end

  # Create the root logger
  # Every other logger will be using his configuration
  def self.create_root
    return Log4r::Logger[ROOT] if Log4r::Logger[ROOT]

    Log4r::Logger.new(ROOT)
    Log4r::Logger[ROOT].additive = false

    otp = Log4r::StdoutOutputter.new('stdout')
    otp.formatter = Log4r::PatternFormatter.new(
      :date_pattern => '%d-%m-%Y %H:%M:%S',
      :pattern => '%l [%d] [%C]: %M')

    Log4r::Logger[ROOT].outputters << otp
    Log4r::Logger[ROOT]
  end
end

File: python/lib/xi/ml/common/logger.py
# encoding: utf-8


import sys
import logging


# Module: handle the predefined and updatable loggers

ROOT = 'xi.ml'
PATH = '.'
FNAME = 'python-xi-ml'
loggers = {}

def create(name):
    """Create a new logger with the class name and return it"""

    if not name.startswith(ROOT):
        return None

    if ROOT not in loggers:
        create_root()

    if name in loggers:
        return loggers[name]

    clogger = logging.getLogger(name)
    clogger.setLevel(loggers.get(ROOT).getEffectiveLevel())
    clogger.handlers = list(loggers.get(ROOT).handlers)
    clogger.propagate = False

    # add to dictionary
    loggers[name] = clogger

    return clogger

def global_level(level):
    """Change the logging level of all existing loggers"""

    for clogger in loggers.values():
        clogger.setLevel(level)

def copy_config(rlogger):
    """Copy the configuration of reference logger onto all existing loggers"""

    if not isinstance(rlogger, logging.Logger):
        return

    for clogger in loggers:
        clogger.setLevel(rlogger.level)
        clogger.handlers = list(rlogger.handlers)


def create_root():
    """Create the root logger: every logger will be using his configuration"""

    # create root if it does not already exist
    if ROOT not in loggers:

        # log formatter
        log_formatter = logging.Formatter(
            '[%(name)s] [%(asctime)s] %(levelname)s : %(message)s')

        # file handler - if needed
        file_handler = logging.FileHandler(
            "{0}/{1}.log".format(PATH, FNAME))
        file_handler.setFormatter(log_formatter)

        # console handler
        console_handler = logging.StreamHandler(sys.stdout)
        console_handler.setFormatter(log_formatter)

        root_logger = logging.getLogger(ROOT)
        root_logger.__name__ = ROOT
        root_logger.addHandler(console_handler)
        root_logger.setLevel(logging.INFO)
        root_logger.propagate = False

        # add to dictionary
        loggers[ROOT] = root_logger

    return loggers[ROOT]

# create ROOT logger
create_root()
